agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+            .lock()
+            .unwrap()
+            .get_mut(&self.config.spec.protocol.name)
+        {
+            if let Some(dh_details) = registered_dh_map.get(endpoint) {
+                let mut dh_details = dh_details.clone();
+                dh_details.connectivity_status = connectivity_status;
+                registered_dh_map.insert(endpoint.to_string(), dh_details);
+            }
+        }
+    }
+
+    /// This is called when no connection can be made with a discovery handler at its endpoint. The discovery handler should be marked offline or removed
+    /// from the RegisteredDiscoveryHandlersMap if it has timed out or its grace period has passed. Will never be
+    pub async fn mark_offline_or_deregister_discovery_handler(
+        &self,
+        endpoint: &str,
+    ) -> Result<bool, Box<dyn std::error::Error + Send + Sync + 'static>> {
+        trace!("mark_offline_or_deregister_discovery_handler - discovery handler at endpoint {} and protocol {} is offline", endpoint, self.config.spec.protocol.name);
+        let mut deregistered = false;
+        if let Some(registered_dh_map) = self
+            .discovery_handler_map
+            .lock()
+            .unwrap()
+            .get_mut(&self.config.spec.protocol.name)
+        {
+            if let Some(dh_details) = registered_dh_map.get(endpoint) {
+                match dh_details.connectivity_status {
+                    DiscoveryHandlerConnectivityStatus::Offline(instant) => {
+                        if instant.elapsed().as_secs() > DISCOVERY_HANDLER_OFFLINE_GRACE_PERIOD_SECS
+                        {
+                            trace!("mark_offline_or_deregister_discovery_handler - de-registering discovery handler for protocol {} at endpoint {} since been offline for longer than 5 minutes", self.config.spec.protocol.name, endpoint);
+                            // Remove discovery handler from map if timed out
+                            registered_dh_map.remove(endpoint);
+                            deregistered = true;
+                        }
+                    }
+                    DiscoveryHandlerConnectivityStatus::Online => {
+                        let mut dh_details = dh_details.clone();
+                        dh_details.connectivity_status =
+                            DiscoveryHandlerConnectivityStatus::Offline(Instant::now());
+                        registered_dh_map.insert(endpoint.to_string(), dh_details);
+                    }
+                    DiscoveryHandlerConnectivityStatus::HasClient => {}
+                }
+            }
+        }
+        Ok(deregistered)
+    }
+
+    /// Checks if any of this DiscoveryOperator's Configuration's Instances have been offline for too long.
+    /// If a non-local device has not come back online before `SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS`,
+    /// the associated device plugin and instance are terminated and deleted.
+    #[allow(dead_code)]
+    pub async fn check_offline_status(
```

Comment:
`check_offline_status` doesn't seem like what this is doing.  it seems like it is `delete_offline_instances`

Reply:
Good point. fixed in 2f6ad02.